Initial commit to check for valid address

eswarasai opened the pull request:
#210

eswarasai commented:
@axic -- So I've added a check for `isValidAddress` and in case of error returning a `modalDialog` with message `Invalid address.`
I can also add an optional check for `isValidChecksumAddress`, but I was wondering what would this check do in case it's valid checksum address or not?

axic commented:
Great!

I think it should pop up a warning if there is a checksum mismatch. There could also be a setting to make it an error.

@yann300 what do you think?

axic commented on src/app/tabs/run-tab.js:
```diff
@@ -10,6 +10,7 @@ var executionContext = require('../../execution-context')
 var copyToClipboard = require('../ui/copy-to-clipboard')
 var Recorder = require('../../recorder')
 var EventManager = require('remix-lib').EventManager
+var ethereumjsUtil = require('ethereumjs-util')
```
I think the usual name for this is `ethUtil`, but please search for other `require('ethereumjs-util')` statements and follow the naming fomr there.

eswarasai replied:
@axic -- There are 2 naming conventions : `ethUtil` and `ethJSUtil`. I'm changing it to `ethJSUtil` as it's being used in majority of the places

axic commented:
Also tests should be added I think.

eswarasai commented:
@axic -- I'm not sure about the tests. If either you or @yann300 can guide me through it, I can take care of it as well.

> I think it should pop up a warning if there is a checksum mismatch

Popup a warning how exactly? Let me know if there's a function already that I can re-use here

yann300 commented:
The test should be added here: https://github.com/ethereum/browser-solidity/blob/master/test-browser/tests/ballot.js#L76

This function https://github.com/ethereum/browser-solidity/blob/master/test-browser/helpers/contracts.js
might needs a bit of modification to check that the alert for `invalid address` is shown.

eswarasai commented:
@yann300 -- Sure. I'll get to it and update you once done

eswarasai commented:
@yann300 -- So I've tried setting up the selenium tests locally by following the steps. But i'm unable to due to the following error :
I execute `npm run test-browser` after doing the `npm run selenium-install`. While executing this command -- `node ci/makeMockCompiler.js`, it's giving me an error saying that -- `Error: Cannot find module '../soljson'`. I tried to download this from `https://ethereum.github.io/solc-bin/soljson.js` and paste it at `browser-solidity/` and then re-execute the command. It runs, tries to download the file -- `https://ethereum.github.io/solc-bin/soljson.js` which it fails to.

Any help here?

yann300 commented:
this command fails ? `npm run downloadsolc`

eswarasai commented:
@yann300 -- Werid. This command works `npm run downloadsolc`. But if I run `npm run test-browser` right after that, it again gives me the same error : `Error: Cannot find module '../soljson'`

yann300 commented:
try runnning:
```
npm run downloadsolc
npm run make-mock-compiler
npm run nightwatch_local
```

eswarasai commented:
@yann300 -- `npm run make-mock-compiler` is throwing the below error : 
```
> browser-solidity@0.0.0 make-mock-compiler /Users/Eswara/Sites/browser-solidity
> node ci/makeMockCompiler.js

TypeError: Cannot read property 'contents' of undefined
    at Array.<anonymous> (/Users/Eswara/Sites/browser-solidity/node_modules/solc/wrapper.js:24:27)
    at jsCall_viii (/Users/Eswara/Sites/browser-solidity/soljson.js:1:908039)
    at Array.M2a (/Users/Eswara/Sites/browser-solidity/soljson.js:9:735166)
    at Array.kd (/Users/Eswara/Sites/browser-solidity/soljson.js:7:31414)
    at Object.L2a [as dynCall_viii] (/Users/Eswara/Sites/browser-solidity/soljson.js:9:735114)
    at invoke_viii (/Users/Eswara/Sites/browser-solidity/soljson.js:1:907877)
    at Array.th (/Users/Eswara/Sites/browser-solidity/soljson.js:7:234585)
    at Object.n4a [as dynCall_viiii] (/Users/Eswara/Sites/browser-solidity/soljson.js:9:741280)
    at invoke_viiii (/Users/Eswara/Sites/browser-solidity/soljson.js:1:908915)
    at sh (/Users/Eswara/Sites/browser-solidity/soljson.js:7:229844)
TypeError: Cannot read property 'contents' of undefined
```

yann300 commented:
should be ok still
and `npm run nightwatch_local`?

eswarasai commented:
@yann300 -- Yep. It's working now. Thanks for the quick response :)

eswarasai commented:
@axic @yann300 -- Can you let me know how should I go about the below query?

> I think it should pop up a warning if there is a checksum mismatch

Popup a warning how exactly? Let me know if there's a function already that I can re-use here

yann300 commented:
you can use it at the same place as the `isValidAddress` is used

eswarasai commented:
@yann300 -- Right now, if the `address` is invalid, I'm displaying that in a `modalDialog`. So I should do the same incase of checksum mismatch as well? Because it's more like an error rather than warning.

yann300 commented:
when you are in both case, you should display it i a `modalDialog.alert` and don't update the runtab

eswarasai commented:
@axic @yann300 -- I've added the alert message in case of invalid checksum address as well. For the tests, I've just updated the `address` passed to the `addInstance()` to a checksum address.
Let me know if there's anything else pending. Thanks.

yann300 commented:
could you add a little change please?
here
https://github.com/ethereum/browser-solidity/pull/1025/files#diff-b2d36c7664b525607de0fb789d7a689cR482
proceed the `isValidChecksumAddress` only if the address contains mixed lower-case, upper-case character.

yann300 commented:
and here : https://github.com/ethereum/browser-solidity/blob/master/test-browser/helpers/contracts.js#L165
There should be a check to fail the test if the wrong modal dialog is displayed.
e.g the input address does not match the requirement (invalid checksum) & the "invalid checksum" modal is not displayed

eswarasai commented:
@yann300 -- Sure. I'll update the code with the above comments. Thanks for the feedback

eswarasai commented:
@yann300 -- I've updated the code with the above comments. Let me know if there's anything else. Thanks

yann300 commented on test-browser/tests/ballot.js:
```diff
@@ -72,11 +72,17 @@ function runTests (browser, testData) {
       })
       .perform((client, done) => {
-        console.log('addInstance 0x692a70d2e424a56d2c6c27aa97d1a86395877b3a')
+        console.log('addInstance invalid checksum address 0x692a70d2e424a56d2c6c27aa97d1a86395877b3a')
```
that address `0x692a70d2e424a56d2c6c27aa97d1a86395877b3a` should not trigger `invalid check sum address` (there's only lowercase)

eswarasai replied:
@yann300 -- That address was just an example. We can also use the combination of uppercase and lowercase letters in order mimic an invalid checksum address

yann300 commented on test-browser/helpers/contracts.js:
```diff
@@ -162,6 +163,9 @@ function addInstance (browser, address, callback) {
     browser.click('div[class^="atAddress"]')
       .perform((client, done) => {
         browser.execute(function () {
+          if (!ethJSUtil.isValidChecksumAddress(address)) {
```
I think it is better to add params in `addInstance` to describe what you expect.
like `function addInstance (browser, address, isValidFormat, isValidChecksum, callback) { ...`
and test if the ui behaves accordingly

eswarasai replied:
@yann300 -- Sure. If you can let me know the exact requirements for these tests, I'd be able to handle those use cases accordingly by adding those `params`.

yann300 replied:
are yo still working on that @eswarasai? I think everything is fine, you just need to update the above test.

eswarasai replied:
@yann300 -- Yes. Sorry about the delay. I'll do that and update within the next 12 hours. Thanks

eswarasai replied:
@yann300 -- I've pushed the changes as per requested above. I've added 2 more params : `isValidFormat, isValidChecksum` to `addInstance`. I haven't updated the tests, they're as it is for now. Let me know if there's a test needed for invalid address as well. Thanks.

axic commented on src/app/tabs/run-tab.js:
```diff
@@ -312,6 +313,12 @@ function contractDropdown (events, appAPI, appEvents, instanceContainer) {
+    if (!ethJSUtil.isValidAddress(address)) {
+      return modalDialogCustom.alert('Invalid address.')
+    }
+    if (/[a-f]/.test(address) && /[A-F]/.test(address) && !ethJSUtil.isValidChecksumAddress(address)) {
```
You could simplify the regexp into one: `/[a-fA-F]/`

eswarasai replied:
@axic -- We'd want to identify if an address consists both lowercase as well as uppercase characters and then only do the validation of checksum address. In case of the above suggested regex, if an address is comprised of either lowercase or uppercase characters will also be matched for which we wouldn't need the validation of checksum address as per the request by @yann300

yann300 commented:
I 've created a PR in your branch https://github.com/eswarasai/browser-solidity/pull/1 
for a fix:
unfortunately I think `browser.assert` is not available inside `browser.execute`

eswarasai commented:
@yann300 -- I've just merged the changes. Thanks for helping me out here with the `browser.assert`. I remember when I was running the test, it showed the failed result appropriately and continued with the other tests. Not really sure what had changed.

Anyway, let me know if we're good to go ahead and merge this PR as well as close the issue. Thanks.

yann300 commented:
could you rebase please? so that will retrigger tests

eswarasai commented:
@yann300 -- Just rebased and committed the changes

eswarasai commented:
@yann300 @axic -- Any specific reason why the tests are taking so long?

yann300 commented:
actually it never started. Could you commit --amend to retrigger build?

eswarasai commented:
@yann300 -- Can you help me in fixing the assertion fail tests for the above build? Thanks.

yann300 commented:
updated. please also rebase the PR , thanks

eswarasai commented:
@yann300 -- I've rebased and pushed the changes. Looks like the tests are still failing.

```
TEST FAILURE:  1 assertions failed, 81 passed. (3m 35s)

 ✖ ballot

   - Ballot (52.38s)
   Failed [fail]: ('cant parse solidity state' undefined 'Unexpected end of JSON input')  - expected "Unexpected end of JSON input" but got: "cant parse solidity state"
       at Object.<anonymous> (/home/circleci/repo/test-browser/helpers/contracts.js:344:22)
```

yann300 commented:
yep that's something else. let me fix that

yann300 commented:
could you rebase again please? thanks

eswarasai commented:
@yann300 -- Looks like still it's failing at the `addInstance` test

```
addInstance 0x692a70D2e424a56D2C6C27aA97D1a86395877b3A
delegate - transact (not payable)
 ✖ Timed out while waiting for element <.instance button[title="delegate - transact (not payable)"]> to be present for 10000 milliseconds.  - expected "found" but got: "not found"
    at Object.testFunction (/home/circleci/repo/test-browser/helpers/contracts.js:120:8)
    at F.browser.waitForElementVisible.click.perform.click.setValue.click.testFunction.pause.click.pause.click.click.click.pause.perform.pause.perform.perform.click.click.perform.perform.perform.perform (/home/circleci/repo/test-browser/tests/ballot.js:88:17)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
```

yann300 commented:
@eswarasai could you rebase please ? test should be ok now
